Added functionality to retry in Invoke-RestMethod and Invoke-WebRequest.

adityapatwardhan opened the pull request:
Fixes #5582

## PR Summary
The change added two parameters, `RetryCount` and `RetryIntervalSec` to enable retry functionality.
When retrying a verbose message is sent out to inform the user.

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [X] Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [X] Issue filed - Issue link: https://github.com/PowerShell/PowerShell-Docs/issues/1991
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [X] [Add `[feature]` if the change is significant or affectes feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

adityapatwardhan committed:
[Feature] Added retry functionality for WebCmdlets

adityapatwardhan committed:
[Feature] Fixed comment

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -229,6 +230,20 @@ public virtual int MaximumRedirection
         private int _maximumRedirection = -1;
 
+        /// <summary>
+        /// gets or sets the RetryCount property
```
Please add a more meaningful summary. Capitalize the first work and end with a period.

adityapatwardhan replied:
Fixed.

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -229,6 +230,20 @@ public virtual int MaximumRedirection
+        public virtual int RetryCount { get; set; } = 0;
+
+        /// <summary>
+        /// gets or sets the RetryIntervalSec property
```
Please add a more meaningful summary. Capitalize the first work and end with a period.

adityapatwardhan replied:
Fixed.

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1265,106 +1280,133 @@ protected override void ProcessRecord()
+
+                        bool isSuccessful = true;
+
+                        do
```
Can you investigate collapsing this to a single do/while instead of nested?

adityapatwardhan replied:
Moved logic to GetResponse()

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1265,106 +1280,133 @@ protected override void ProcessRecord()
+
+                                        WriteVerbose(retryMessage);
+                                        Task.Delay(RetryIntervalSec * 1000).GetAwaiter().GetResult();
+                                        continue;
```
We should probably dispose the current HttpResponseMessage before retrying.

adityapatwardhan replied:
fixed

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1265,106 +1280,133 @@ protected override void ProcessRecord()
+                                            WebCmdletStrings.RetryVerboseMsg,
+                                            RetryIntervalSec,
+                                            response.StatusCode
+                                            );
```
This should move to the end of the previous line.

adityapatwardhan replied:
Fixed.

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1265,106 +1280,133 @@ protected override void ProcessRecord()
                         }
 
-                        using (HttpRequestMessage request = GetRequest(uri, stripAuthorization:false))
+                        int retriesLeft = RetryCount;
```
I believe it would be better to have `RetryCount` and `RetryIntervalSec` set on `WebSession` in `PrepareSession()` and then have this area of code retrieve the values from `WebSession`. It would mean extending `WebRequestSession` to include new Properties, but I believe that is the correct direction for these settings.

adityapatwardhan replied:
Fixed.

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1265,106 +1280,133 @@ protected override void ProcessRecord()
+
+                                    isSuccessful = response.IsSuccessStatusCode && !causeFailureForTesting;                                
+
+                                    if (!isSuccessful && retriesLeft > 0)
```
This current logic leads to an interestingly weird combination. If you set `-MaximumRedirection 0 -RetryCount 5 -RetryIntervalSec 1` and the response code is `301`, this will still retry the same link 5 times because `HttpResponseMessage.IsSuccessStatusCode` is only true for response codes `200`-`299` inclusive... not sure how to rectify that.

markekraus replied:
also... since the logic for reducing the current redirection count is handled in `GetResponse()`, we may need to reset `WebSession.MaximumRedirection` to the `MaximumRedirection` property value before continuing.

SteveL-MSFT replied:
Best that I can find, the HTTP spec doesn't say enough about rate limiting to make it something we can depend on reliably.  However, I think we can say that retry only occurs on a 429 or 403 and document it as such unless there's data showing that web services are using other error codes.

markekraus replied:
@SteveL-MSFT I can see other usecases for this beyond just rate limiting. I could see using this to test 404 errors until a new page comes online after a publish/push and it $x attempts fail alert that the push/publish failed. 

Perhaps it we limit it to 400-499 inclusive it would be best. In my experience with APIs I have seen more than 403 and 429 used for busy endpoints (including 404... Reddit... ugh) or rate limiting, but it's always a 400 level error

SteveL-MSFT replied:
@markekraus that's a fair point.  I think 4xx would be fine.

markekraus replied:
After further consideration, this should work for the following status codes: 304. 400-599 inclusive

304 could be used for monitoring the change of a resource and is not a redirect status even though it in the 300 range. Retries on 5XX errors makes perfect sense to overcome temporary service glitches and cloudflare shenanigans.

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1265,106 +1280,133 @@ protected override void ProcessRecord()
+                                            );
+
+                                        WriteVerbose(retryMessage);
+                                        Task.Delay(RetryIntervalSec * 1000).GetAwaiter().GetResult();
```
See my comment about setting and retrieving `RetryIntervalSec` and `RetryCount` from `WebSession`.

adityapatwardhan replied:
Fixed.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -825,6 +825,29 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+            $result.output.StatusCode | Should Be "200"
+
+            # Expect 2 verbose messages as we are retrying twice.
+            (Select-String -Path $outputFilePath -Pattern 'Retrying after interval').Count | Should Be 2
```
This won't work once we start localizing the error messages

markekraus replied:
I don't really like this test.  I'd rather have a real failure emulated by `WebListener`. Add a new endpoint which take a session ID (new GUID), a status code, and a fail count. Until the initially provided fail count is reached when using that session ID it produces a response with the provided status code After fail count is reached it returns a normal 200 response. The response body should include the request count for the given session ID. This way we can test how it reacts with various responses and differing fail counts provided by actual endpoint errors instead of relying on fake logic to enter the code path and without relying on verbose messages.

iSazonov commented:
@adityapatwardhan Could you please continue?

iSazonov commented:
@adityapatwardhan Please resolve conflicts.

adityapatwardhan commented:
I plan to resume work on this next week.

ffeldhaus commented:
Did you consider implementing Exponential Backof or different retry strategies? [AWS SDKs for instance use an Exponential Backof](https://docs.aws.amazon.com/general/latest/gr/api-retries.html) and it is good practice for many other web services.

iSazonov commented:
@ffeldhaus Please open new Issue for your great suggestion - we shouldn't lost this.

ffeldhaus commented:
I thought about this feature some more and I don't think it should be included in the current form without further discussion to prevent retries to cause issues or unwanted behavior and I don't think a Pull Request is the right place to discuss the design of this feature. As issue #5582 was already created for this feature, we should first discuss the design there and then continue with the Pull Request.

I will add my concerns and design ideas to issue #5582.

markekraus commented:
@iSazonov Retry and Resume are separate features. Resume continues downloading a partially downloaded file. Retry tries the URI again if a non-susccess status code was received. This PR is for the Retry feature, #6447 is for the Resume feature.

iSazonov commented:
I wouldn't want to lose this.

adityapatwardhan committed:
Move retry logic to GetResponse

adityapatwardhan commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -229,6 +230,20 @@ public virtual int MaximumRedirection
+        /// </summary>
+        [Parameter]
+        [ValidateRange(1, Int32.MaxValue)]
+        public virtual int RetryIntervalSec { get; set; } = 10;
```
Fixed.

adityapatwardhan committed:
Move retry logic to GetResponse and address code review feedback

adityapatwardhan committed:
Add retry controller in WebListener

adityapatwardhan committed:
Fix bug in number of retries

adityapatwardhan committed:
Add response to retry controller

adityapatwardhan committed:
Fix retry loop bug

adityapatwardhan committed:
[Feature] Updated tests

adityapatwardhan commented:
@iSazonov @markekraus @SteveL-MSFT  I have addressed all the feedback. Please have another look.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1268,80 +1290,117 @@ static bool IsRedirectToGet(HttpStatusCode code)
 
+        bool ShouldRetry(HttpStatusCode code)
+        {
+            int intCode = (int) code;
```
Please remove extra space after `(int)`. (See CodeFactor issues.)

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -630,6 +645,14 @@ internal virtual void PrepareSession()
                 }
             }
+
+            if(RetryCount > 0)
```
Please add space after `if`. (See CodeFactor issues.))

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1268,80 +1290,117 @@ static bool IsRedirectToGet(HttpStatusCode code)
+                _resumeSuccess = response.StatusCode == HttpStatusCode.PartialContent;
 
-                using (HttpRequestMessage requestWithoutRange = GetRequest(currentUri))
+                if(ShouldRetry(response.StatusCode))
```
Please add space after `if`. (See CodeFactor issues.)

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -228,6 +229,20 @@ public virtual int MaximumRedirection
         private int _maximumRedirection = -1;
 
+        /// <summary>
+        /// Gets or Sets the RetryCount property, which determines the number of retries of a failed web request.
```
CodeFactor wants  "Gets or sets".

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -228,6 +229,20 @@ public virtual int MaximumRedirection
+        public virtual int RetryCount { get; set; } = 0;
+
+        /// <summary>
+        /// Gets or Sets the RetryIntervalSec property, which determines the number seconds between retries.
```
CodeFactor wants  "Gets or sets".

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -64,6 +64,16 @@ public class WebRequestSession
+        public int RetryCount { get; set; }
+
+        /// <summary>
+        /// Gets or Sets the interval in seconds between retries.
```
CodeFactor wants  "Gets or sets".

iSazonov commented on test/tools/WebListener/Startup.cs:
```diff
@@ -79,6 +79,10 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
+                routes.MapRoute(
+                    name: "retry",
+                    template: "Retry/{sessionId?}/{failureCode?}/{failureCount?}",
+                    defaults: new {controller = "Retry", action = "Retry"});
```
CodeFactor wants spaces with braces.
https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1012.md

iSazonov commented on test/tools/WebListener/Controllers/RetryController.cs:
```diff
@@ -0,0 +1,64 @@
+                var newRetryInfoItem = Tuple.Create(failureCode, failureCount, 1);
+                retryInfo.Add(sessionId, newRetryInfoItem);
+                Response.StatusCode = failureCode;
+                Hashtable error = new Hashtable {{"error", $"Error: HTTP - {failureCode} occurred."}};
```
CodeFactor wants spaces with braces.

iSazonov commented on test/tools/WebListener/Controllers/RetryController.cs:
```diff
@@ -0,0 +1,64 @@
+            }
+            else
+            {
+                //initialize the failureResponsesSent as 1.
```
Please add space after `//`

iSazonov commented on test/tools/WebListener/Controllers/RetryController.cs:
```diff
@@ -0,0 +1,64 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Http.Extensions;
```
Wrong sorting - please swap lines 11 and 12.

iSazonov commented on test/tools/WebListener/Controllers/RetryController.cs:
```diff
@@ -0,0 +1,64 @@
+
+        public JsonResult Retry(string sessionId, int failureCode, int failureCount)
+        {
+            if(retryInfo == null)
```
Please add space after `if`.

iSazonov commented on test/tools/WebListener/Controllers/RetryController.cs:
```diff
@@ -0,0 +1,64 @@
+                retryInfo = new Dictionary<string, Tuple<int, int, int>>();
+            }
+
+            if(retryInfo.TryGetValue(sessionId, out Tuple<int, int, int> retry))
```
Please add space after `if`.

iSazonov commented on test/tools/WebListener/Controllers/RetryController.cs:
```diff
@@ -0,0 +1,64 @@
+            if(retryInfo.TryGetValue(sessionId, out Tuple<int, int, int> retry))
+            {
+                // if failureResponsesSent is less than failureCount
+                if(retry.Item3 < retry.Item2)
```
Please add space after `if`.

iSazonov commented on test/tools/WebListener/Controllers/RetryController.cs:
```diff
@@ -0,0 +1,64 @@
+        }
+
+        public IActionResult Error()
+
```
Please remove extra line.

iSazonov commented on test/tools/WebListener/Controllers/RetryController.cs:
```diff
@@ -0,0 +1,64 @@
+using Microsoft.Extensions.Primitives;
+using mvc.Models;
+
+namespace mvc.Controllers
```
CodeFactor - Namespace names begin with an upper-case letter
https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1300.md

adityapatwardhan replied:
This will cause unnecessary changes across the WebListener project. I prefer not fixing this.

iSazonov reviewed (commented):
My comments come from CodeFactor - I believe we should fix the issues if we have changed the code.

adityapatwardhan committed:
Address CodeFactor comments

adityapatwardhan committed:
[Feature] Address feedback about tests

adityapatwardhan commented:
@iSazonov code review comments addressed.

adityapatwardhan committed:
Add WebListener documenation

adityapatwardhan committed:
[Feature] CodeFactor fixes

adityapatwardhan commented:
@markekraus Added documentation and fixed more CodeFactor issues.

adityapatwardhan commented:
@SteveL-MSFT  Can you have another look?

iSazonov commented:
Restarted Travis CI Linux - seems temporary failed.

adityapatwardhan committed:
Add to spellings file

adityapatwardhan commented:
@SteveL-MSFT  @markekraus  Please update your reviews.

markekraus commented:
@adityapatwardhan I may not have time until this weekend to give this the attention it deserves. I do want to review it for edge cases before it's pulled.

ffeldhaus commented:
Is it really ok to do the retry for Status Code 304 and 400 to 599? There are several Status Codes were a retry just does not make sense and will just increase the load/error counters on the server. 400, 409, 410, 429, 431, 505 would be some were I don't think a retry should be done. 413 would require special retry handling.

markekraus commented:
@ffeldhaus I have plenty of use cases for retries just about every status code. For example, the status code is server side bug, a fix is being released to prod, and I want the script to retry until x amount of time passes or the fix is live.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs:
```diff
@@ -64,6 +64,16 @@ public class WebRequestSession
+        /// <summary>
+        /// Gets or sets the count of retries for request failures.
+        /// </summary>
+        public int RetryCount { get; set; }
```
To be consistent, I wonder if this should be `MaximumRetry`

adityapatwardhan replied:
I like this suggestion, I will make this change.

markekraus replied:
`MaximumRetry` sounds weird. Maybe `MaximumRetryCount`?

Looking at existing commands, `New-PSSessionOption` uses `-MaxConnectionRetryCount`

adityapatwardhan replied:
Fixed.

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1268,80 +1290,121 @@ static bool IsRedirectToGet(HttpStatusCode code)
+                    WriteVerbose(retryMessage);
+                    Task.Delay(WebSession.RetryIntervalInSeconds * 1000).GetAwaiter().GetResult();
+                    req.Dispose();
+                    req = GetRequest(Uri);
```
`Uri` should be `currentUri`

markekraus replied:
`FillRequestStream(req)` needs to called to ensure the body is present.

There should also be a test to ensure body data survives the retries.

```powershell
Invoke-WebRequest -Uri $Uri -Body 'a' -Method 'POST' -RetryCount 1  -RetryIntervalSec 1
```

with fiddler the body on the first request is correct. the body on the second request is empty.

adityapatwardhan replied:
Good catch. Thanks.

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1268,80 +1290,121 @@ static bool IsRedirectToGet(HttpStatusCode code)
+
+                _resumeSuccess = response.StatusCode == HttpStatusCode.PartialContent;
+
+                if (ShouldRetry(response.StatusCode))
```
The logic here results in the verbose message printed an extra time when the final result is still an error:

```powershell
Invoke-RestMethod -RetryCount 1 -RetryIntervalSec 1 https://httpbin.org/status/404 -Verbose -SkipCertificateCheck
```

```none
VERBOSE: GET https://httpbin.org/status/404 with 0-byte payload
VERBOSE: Retrying after interval of 1 seconds. Status code for previous attempt: NotFound
VERBOSE: Retrying after interval of 1 seconds. Status code for previous attempt: NotFound
VERBOSE: received 0-byte response of content type text/html
irm : Response status code does not indicate success: 404 (NOT FOUND).
At line:1 char:1
+ irm -RetryCount 1 -RetryIntervalSec 1 https://httpbin.org/status/404  ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (Method: GET, Re...rShell/6.1.0
}:HttpRequestMessage) [Invoke-RestMethod], HttpResponseException
+ FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand
```

Maybe `if (totalRequests<0 && ShouldRetry(response.StatusCode))`?

adityapatwardhan replied:
Fixed.

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1268,80 +1290,121 @@ static bool IsRedirectToGet(HttpStatusCode code)
+                        response.StatusCode);
+
+                    WriteVerbose(retryMessage);
+                    Task.Delay(WebSession.RetryIntervalInSeconds * 1000).GetAwaiter().GetResult();
```
is there some better way to do this that will be responsive to ctrl+alt+del? Currently, if you use the default 60 seconds, got an error endpoint without verbose it looks like powershell is hung because ctrl+alt+delete doesn't end the command.

adityapatwardhan replied:
Added a cancellation token for the Delay. Ctrl+C will cancel the delay and the webrequest.

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -228,6 +229,20 @@ public virtual int MaximumRedirection
+        /// </summary>
+        [Parameter]
+        [ValidateRange(1, Int32.MaxValue)]
+        public virtual int RetryIntervalSec { get; set; } = 60;
```
60 seconds as the default just feels too long. So many times testing this feature and forgetting the `-RetryIntervalSec` I felt like the console froze or the thread locked. between 10 and 15 seconds I can't decide which is better. 10 feels more natural, but 15 is probably better. anything over 15 really starts to feel like the command has frozen.

adityapatwardhan replied:
According to Azure docs [here](https://docs.microsoft.com/en-us/aspnet/aspnet/overview/developing-apps-with-windows-azure/building-real-world-cloud-apps-with-windows-azure/transient-fault-handling#use-smart-retryback-off-logic-to-mitigate-the-effect-of-transient-failures) the sample shows 5 seconds. @markekraus  @SteveL-MSFT what do you think?

markekraus replied:
I would agree with 5 seconds for transient errors, but we are covering more than transient errors (for example, 404). But... maybe it's ok. 5 seconds is still only 12 requests a minute which isn't that spammy, IMO.

adityapatwardhan replied:
Changing it to 5 seconds, if we get feedback about it to increase the defaults, we can reconsider.

adityapatwardhan committed:
[Feature] Address code review feedback

markekraus commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs:
```diff
@@ -1268,80 +1290,128 @@ static bool IsRedirectToGet(HttpStatusCode code)
+
+                    req.Dispose();
+                    req = GetRequest(currentUri);
+                    FillRequestStream(req);
```
@adityapatwardhan Can you either add a test for retry POST data or open an issue to have one added?

adityapatwardhan replied:
Added a test for `POST`

adityapatwardhan committed:
[Feature] Added test for 'POST'

adityapatwardhan commented:
@markekraus Added a test for POST, please have another look. Thanks!

markekraus commented:
@adityapatwardhan That test doesn't provide a body payload nor does it test the response. The WebListener endpoint would need to echo back the body payload and then the pester test would need to check the response from WebListener to ensure the echoed payload matches the sent payload.

adityapatwardhan committed:
[Feature] Added sessionId echo for RetryController and updated test

adityapatwardhan commented:
@markekraus Updated test and controller with echo for sessionId.

markekraus commented:
@adityapatwardhan Two more things and I think this will be done. I just realized there are only tests for `Invoke-Webrequest`. Please add the same tests for `Invoke-RestMethod`. Also, instead of wrapping the tests in `#region Retry tests`, please use a `Context` block.

adityapatwardhan commented:
@markekraus Since the code path is same, I believe adding all the tests will be duplication. I will add one test for Invoke-RestMethod to cover the scenario.

markekraus commented:
@adityapatwardhan The convention for the web cmdlets is to test all features on both, even common features. There is always the possibility that the common features clash with the implementation features.

adityapatwardhan committed:
[Feature] Added tests for Invoke-RestMethod

adityapatwardhan commented:
@markekraus Please review again. I have added tests for `Invoke-RestMethod`. I moved them to a new describe block so they can be grouped together. The describes for other Invoke-RestMethod and Invoke-WebRequest tests were separate. Thanks for the feedback, very much appreciated.

markekraus commented:
@adityapatwardhan there is a `Describe "Invoke-WebRequest tests" -Tags "Feature"` block and a `Describe "Invoke-RestMethod tests" -Tags "Feature"` and in those are various context blocks 

for example, 

https://github.com/PowerShell/PowerShell/blob/e8d2a6b27827dd366f638468f96487ddf2de1db4/test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1#L827

and 

https://github.com/PowerShell/PowerShell/blob/e8d2a6b27827dd366f638468f96487ddf2de1db4/test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1#L2119

adityapatwardhan commented:
@markekraus I decided to go with a separate `Describe` block so that we can use `TestCase` for `It` and re-use most of the test code. If we have tests for `Invoke-RestMethod` and `Invoke-WebRequest` in separate `Describe` blocks then there will be test code duplication. I prefer a separate `Describe` but since there is precedence I am not completely against adding them to existing `Describe` blocks.

Let me know your thoughts?

markekraus commented:
@adityapatwardhan For all new features they have been separate context blocks in the top level describe blocks for each cmdlet. I'd prefer to stick to this pattern for now and look at changing things up in one of the open issues I have on refactoring the code and tests.

adityapatwardhan commented:
@markekraus Ok, agreed. I will send an update in a few minutes.

adityapatwardhan committed:
[Feature] Moved tests in existing Describe blocks

adityapatwardhan commented:
@markekraus Made the changes to the tests as discussed above.

markekraus commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -1723,6 +1723,49 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+
+    Context "Invoke-WebRequest retry tests" {
+
+        It "<Command> can retry - <Name>" -TestCases @(
```
`<Command>` -> `Invoke-WebRequest` since it's no longer in the testcases hash.

markekraus commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -1723,6 +1723,49 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+            $jsonError.error | Should -BeExactly 'Error: HTTP - 400 occurred.'
+        }
+
+        It "<Command> can retry with POST" {
```
`<Command>` -> `Invoke-WebRequest` since it's no longer in the testcases hash.

adityapatwardhan committed:
[Feature] Fixed test title

adityapatwardhan commented:
@markekraus  Good catch! Fixed.

markekraus commented on test/tools/WebListener/README.md:
````diff
@@ -682,3 +682,20 @@ X-WebListener-Has-Range: false
+
+```json
+{
+  "failureResponsesSent":2
````
Can you update this to include the `sessionId` that is now also returned.

markekraus reviewed (changes requested):
One final thing. Sorry :(

adityapatwardhan committed:
Fix documentation

adityapatwardhan commented:
@markekraus  Fixed documentation.

markekraus reviewed (approved):
@adityapatwardhan Thanks for you patience and diligence.  LGTM

iSazonov commented:
Reopen to restart CIs.

adityapatwardhan commented:
@iSazonov  Thanks for restarting CIs. All tests passed, ready to merge?

iSazonov commented:
@adityapatwardhan Thanks for great contribution!
